### PR TITLE
FEVM support

### PR DESCRIPTION
**Pull request by cypt4**
Resolves https://github.com/brave/brave-browser/issues/28954

Etherium addrs supported on filecoin network using new Filecoin EVM tokens

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

At the moment only calibration network support etherium-like addresses.
You can get test filecoins from calibration faucet https://calibration-faucet.filswan.com/ on your eth account
by converting it's address by https://beryx.zondax.ch/address_converter and switching first letter f -> t
New tokens related to Filecoin EVM should be added.
Sending\Opening blockscan\Hiding\Showing should be tested.
Also it is important to test for an update from previous version - new tokens should be properly added

### CODE REVIEW

**Review by Douglashdaniel — APPROVED** [review]
++ Front-end!

File: browser/brave_wallet/brave_wallet_prefs_unittest.cc [hunk 1]
```diff
@@ -232,7 +232,8 @@ TEST_F(BraveWalletPrefsUnitTest, MigrateShowTestNetworksToggle) {
   // Test networks are hidden by default.
   EXPECT_THAT(GetHiddenNetworks(GetPrefs(), mojom::CoinType::ETH),
               ElementsAreArray({mojom::kGoerliChainId, mojom::kSepoliaChainId,
-                                mojom::kLocalhostChainId}));
+                                mojom::kLocalhostChainId,
+                                mojom::kFilecoinEtheriumTestnetChainId}));
```
Comment (yrliou) [inline browser/brave_wallet/brave_wallet_prefs_unittest.cc:236]: s/Etherium/Ethereum in all places

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 2]
```diff
@@ -864,6 +864,64 @@ void BraveWalletService::MigrateUserAssetsAddIsNFT(PrefService* prefs) {
...
+      base::Value::List* list = dict->EnsureList(filecoin_evm_testnet);
+      auto value = base::Value(GetEthNativeAssetFromChain(
+          GetKnownChain(nullptr, filecoin_evm_testnet, mojom::CoinType::ETH)));
+      if (std::find_if(list->begin(), list->end(), [&value](auto& v) {
+            return value == v;
+          }) == list->end()) {
+        list->Insert(list->begin(), std::move(value));
+      }
+    }
+  }
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_service.cc:906]: Can we migrate this part using existing MigrateUserAssetsAddPreloadingNetworks?
See https://github.com/brave/brave-core/pull/14080/files where we added another network and just updating kBraveWalletUserAssetsAddPreloadingNetworksMigrated.

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 3]
```diff
@@ -864,6 +864,64 @@ void BraveWalletService::MigrateUserAssetsAddIsNFT(PrefService* prefs) {
...
+    auto& hidden_networks_pref = update.Get();
+    base::Value::List* hidden_eth_networks =
+        hidden_networks_pref.FindList(kEthereumPrefKey);
+    auto value = base::Value(mojom::kFilecoinEtheriumTestnetChainId);
+    if (std::find_if(hidden_eth_networks->begin(), hidden_eth_networks->end(),
+                     [&value](auto& v) { return value == v; }) ==
+        hidden_eth_networks->end()) {
+      hidden_eth_networks->Append(std::move(value));
+    }
+  }
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_service.cc:920]: Could we make this part generalized and have a function named similar as `MigrateHiddenNetworksAddPreloadingTestnets`, and re-used it every time we're adding a new test network?

File: components/brave_wallet/browser/pref_names.cc [hunk 4]
```diff
@@ -36,9 +36,11 @@ const char kBraveWalletUserAssets[] = "brave.wallet.wallet_user_assets";
 const char kBraveWalletUserAssetEthContractAddressMigrated[] =
     "brave.wallet.user.asset.eth_contract_address_migrated";
 const char kBraveWalletUserAssetsAddPreloadingNetworksMigrated[] =
-    "brave.wallet.user.assets.add_preloading_networks_migrated_2";
+    "brave.wallet.user.assets.add_preloading_networks_migrated_3";
```
Comment (yrliou) [inline components/brave_wallet/browser/pref_names.cc:39]: Please added `kBraveWalletUserAssetsAddPreloadingNetworksMigrated2` under DEPRECATED section, and please add `ClearPref` calls for `kBraveWalletUserAssetsAddPreloadingNetworksMigratedDeprecated` and `kBraveWalletUserAssetsAddPreloadingNetworksMigratedDeprecated2` at the end of `MigrateUserAssetsAddPreloadingNetworks` so we don't store these deprecated prefs.

File: components/brave_wallet/browser/brave_wallet_service.cc [hunk 5]
```diff
@@ -864,6 +864,26 @@ void BraveWalletService::MigrateUserAssetsAddIsNFT(PrefService* prefs) {
   prefs->SetBoolean(kBraveWalletUserAssetsAddIsNFTMigrated, true);
 }
 
+// static
+void BraveWalletService::MigrateHiddenNetworks(PrefService* prefs) {
+  auto previous_version_code =
+      prefs->GetInteger(kBraveWalletDefaultHiddenNetworksVersion);
+  if (previous_version_code < 1) {
```
Comment (yrliou) [inline components/brave_wallet/browser/brave_wallet_service.cc:871]: Return early `if (previous_version_code >= 1)` (latest version that nothing to migrate), `prefs->SetInteger(kBraveWalletDefaultHiddenNetworksVersion, 1);` shouldn't need to be run if this is true.

**Review by yrliou — APPROVED** [review]
LGTM

**Comment by srirambv** [comment]
Verification passed on 

Brave | 1.51.57 Chromium: 112.0.5615.29 (Official Build) nightly (64-bit)
-- | --
Revision | `8a59304c62e066962363d2fd8edd8e7b459c42bd-refs/branch-heads/5615@{#497}`
OS | Linux

- Verified steps from PR 
- Verified upon upgrade to 1.51.57, `Filecoin EVM Mainnet -0x13a https://api.node.glif.io/rpc/v1` &  `Filecoin EVM Testnet - 0x4cb2f https://api.calibration.node.glif.io/rpc/v1` is shown under network list 
- Verified `Filecoin EVM Mainnet` is enabled by default and `Filecoin EVM Testnet` is disabled by default
- Verified able to send `FIL on Filecoin EVM Testnet` from one account to another
- Verified clicking on `View on blockchain explorer` loads the correct block explorer
- Verified sent transaction is listed under activity tab
- Verified selecting `FIL on Filecoin EVM Mainnet/Testnet` shows all the accounts that hold the balance of the token
- Verified able to hide network and hide the token from portfolio page

Transaction | Upgrade scenario
-- | --
<video src=https://user-images.githubusercontent.com/17010094/227448607-9c75d313-a686-4852-8d02-a4cfae8bb5a7.mp4> | <video src=https://user-images.githubusercontent.com/17010094/227448628-90a64219-3b63-4686-9709-41b13c03c609.mp4>

**Pull request merged**